File: main_2.py
import fhirton as ft
import os

if __name__ == '__main__':

    # fhir endpoint
    endp = "https://hapi.fhir.org/baseR4/"

    # fhir request
    req = "Observation?" \
          "code=http://loinc.org|85354-9&" \
          "_include=Observation:patient&" \
          "_include=Observation:encounter&" \
          "_format=xml&_count=500" \

    req = endp + req

    # designs 3, 2, 3
    designs = {
        "Observations": (".//Observation",),
        "Encounters": (".//Encounter", './/*'),
        "Patients": (".//Patient",)
    }

    # download fhir bundles and babble a lot
    bundles = ft.fhir_search(req=req, verbose=2)

    # crack/flatten the downloaded bundles to tables via designs
    # use space as separator and [] as brackets for indexing and babble a lot
    tables = ft.fhir_ton(bundles=bundles, designs=designs, sep=" ", bra=["[", "]"], verbose=3)

    # sort column names
    for k in tables.keys():
        tables[k].sort_index(axis=1, inplace=True)

    # create dir if not exist
    if not ('csv2' in os.listdir(".")):
        os.mkdir("csv2")

    # write tables as csv files
    for k in tables.keys():
        tables[k].to_csv("csv2/" + k + "_python.csv", sep=";", decimal=".", encoding="utf-8", index=False, na_rep='')

File: main_2.R
# Title     : main.R
# Objective : fhirton fhirckrar comparison
# Created by: tpeschel
# Created on: 19.07.20

devtools::install_github("POLAR-fhir/fhircrackr", quiet = T)
#install.packages( "fhircrackr" )
library('dplyr')
library('fhircrackr')

# fhir endpoint
endp <- "https://hapi.fhir.org/baseR4/"

# fhir request
req  <- paste0(
  "Observation?",
  "code=http://loinc.org|85354-9&",
  "_include=Observation:patient&",
  "_include=Observation:encounter&",
  "_format=xml&_pretty=true&_count=500" )

#req <- "Observation?_include=Observation:patient&_include=Observation:encounter&_format=xml&_pretty=true&_count=500"
#req <- "Observation?_include=Observation:patient&_include=Observation:encounter&_format=xml&_pretty=true&_count=500"

fsq <- paste_paths(endp, req)

# designs 3, 2, 3
design <- list(
  "Observations" = list(
    ".//Observation"
  ),
  "Encounters" = list(
    ".//Encounter",
    ".//*"
  ),
  "Patients" = list(
    ".//Patient"
  )
)

# download fhir bundles and babble a lot
bundles <- fhir_search(fsq, max_bundles = 3, verbose=2)

# crack/flatten the downloaded bundles to tables via designs
# use space as separator and [] as brackets for indexing and babble a lot
tables <- fhir_crack(bundles, design, sep = " ", brackets = c('[', ']'), verbose = 2)

# sort column names
tables <- lapply(
  tables,
  function(tbl) {
    tbl[, sort(names(tbl))]
  }
)

# create dir if not exist
if (! dir.exists('csv2'))
  dir.create( 'csv2', recursive = T )

# write tables as csv files
for (n in names(tables))
  write.table(tables[[n]], paste0("csv2/", n, "_r.csv" ), na = "", sep = ";", dec = ".", row.names = F, quote = F )
